ocrd_segment/repair.py

```diff
@@ -87,49 +88,78 @@ def process(self):
 
+            # TODO: cover recursive region structure (but compare only at the same level)
             regions = page.get_TextRegion()
-            for i in range(0, len(regions)):
-                for j in range(i+1, len(regions)):
-                    region1 = regions[i]
-                    region2 = regions[j]
+            # sort by area to ensure to arrive at a total ordering compatible
+            # with the topological sort along containment/equivalence arcs
+            # (so we can avoid substituting regions with superregions that have
+            #  themselves been substituted/deleted):
+            RegionPolygon = namedtuple('RegionPolygon', ['region', 'polygon'])
+            regionspolys = sorted([RegionPolygon(region, Polygon(polygon_from_points(region.get_Coords().points)))
+                                   for region in regions],
+                                  key=lambda x: x.polygon.area)
+            for i in range(0, len(regionspolys)):
+                for j in range(i+1, len(regionspolys)):
+                    region1 = regionspolys[i].region
+                    region2 = regionspolys[j].region
+                    poly1 = regionspolys[i].polygon
+                    poly2 = regionspolys[j].polygon
                     LOG.debug('Comparing regions "%s" and "%s"', region1.id, region2.id)
-                    region1_poly = Polygon(polygon_from_points(region1.get_Coords().points))
-                    region2_poly = Polygon(polygon_from_points(region2.get_Coords().points))
                     
-                    equality = region1_poly.almost_equals(region2_poly)
-                    if equality:
-                        LOG.warning('Page "%s" regions "%s" and "%s" cover the same area.',
-                                    page_id, region1.id, region2.id)
-                        mark_for_deletion.append(region2)
+                    if poly1.almost_equals(poly2):
+                        LOG.warning('Page "%s" region "%s" is almost equal to "%s" %s',
+                                    page_id, region2.id, region1.id,
+                                    '(removing)' if plausibilize else '')
+                        mark_for_deletion.append(region2.id)
+                    elif poly1.contains(poly2):
+                        LOG.warning('Page "%s" region "%s" is within "%s" %s',
+                                    page_id, region2.id, region1.id,
+                                    '(removing)' if plausibilize else '')
+                        mark_for_deletion.append(region2.id)
+                    elif poly2.contains(poly1):
+                        LOG.warning('Page "%s" region "%s" is within "%s" %s',
+                                    page_id, region1.id, region2.id,
+                                    '(removing)' if plausibilize else '')
+                        mark_for_deletion.append(region1.id)
+                    elif poly1.overlaps(poly2):
```

Comment:
Why can't we just have this condition? I.e. the others are special cases of this one.

Reply:
`overlaps()` already excludes `contains()` / `within()`. 

Apart from upholding the distinction between delete and merge (see next answer), I think it is useful to differentiate multiple cases here, because these shapely predicates are much faster than the area test.

Reply:
See next answer.